verification/verify.go

```diff
@@ -0,0 +1,139 @@
+package verification
+
+import (
+	"errors"
+	"sort"
+
+	"github.com/livepeer/go-livepeer/core"
+	"github.com/livepeer/go-livepeer/net"
+
+	"github.com/livepeer/lpms/ffmpeg"
+	"github.com/livepeer/lpms/stream"
+)
+
+// Special error type indicating a retryable error
+// Such errors typically mean re-trying the transcode might help
+// (Non-retryable errors usually indicate unrecoverable system errors)
+type Retryable struct {
+	error
+}
+
+var ErrPixelMismatch = Retryable{errors.New("PixelMismatch")}
+
+type Params struct {
+	// ManifestID should go away once we do direct push of video
+	ManifestID core.ManifestID
+
+	// Bytes of the source video segment
+	Source *stream.HLSSegment
+
+	// Rendition parameters to be checked
+	Profiles []ffmpeg.VideoProfile
+
+	// Information on the orchestrator that performed the transcoding
+	Orchestrator *net.OrchestratorInfo
+
+	// Transcoded result metadata
+	Results *net.TranscodeData
+
+	// Rendition locations; typically when the data is in object storage
+	URIs []string
+
+	// Cached data when local object storage is used
+	Renditions [][]byte
+}
+
+type Results struct {
+	// Verifier specific score
+	Score float64
+
+	// Number of pixels decoded in this result
+	Pixels []int64
+}
+
+type Verifier interface {
+	Verify(params *Params) (*Results, error)
+}
+
+type Policy struct {
+	// Verification function to run
+	Verifier Verifier
+
+	// Maximum number of retries until the policy chooses a winner
+	Retries int
+
+	// How often to invoke the verifier, on a per-segment basis
+	SampleRate float64 // XXX for later
+
+	// How many parallel transcodes to support
+	Redundancy int // XXX for later
+}
+
+type SegmentVerifierResults struct {
+	params *Params
+	res    *Results
+}
+
+type byResScore []SegmentVerifierResults
+
+func (a byResScore) Len() int           { return len(a) }
+func (a byResScore) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byResScore) Less(i, j int) bool { return a[i].res.Score < a[j].res.Score }
```

Comment:
Won't this result in the first element of the sorted slice having the lowest score when we want the first element of the sorted slice to have the highest score?